tests, infra-test, Fix wrong error handling that mask real connectivity errors, and add missing vmi deletion.

oshoval opened the pull request:
Test 4140 checks metrics access throttling
which is expected to raise errors,
but it didn't distinct between a connectivity error
and other errors.
Add a validation that the error is only an expected error
and not due connectivity (network unreachable)

Refactor some tests to use DescribeTable.

Remove unneeded vmi deletion (test 4146).
In order to make test 4146 robust to use
DescribeTable, removed the extra VM creation,
because the test is part of a context which
already have 2 vms as it requires.
This way it would be able to adapted to
DescribeTable without need to clean the machines
each tests, as it count them according the metrics,
and check the reported value against the expected
one.

See: #4145

```release-note
None
```

RamLavi commented on tests/infra_test.go:
```diff
@@ -696,6 +697,9 @@ var _ = Describe("[Serial]Infrastructure", func() {
 				if err != nil {
+					urlErr, ok := err.(*netURL.Error)
+					Expect(ok).To(BeTrue())
+					Expect(urlErr.Err.Error()).To(ContainSubstring("Client.Timeout exceeded while awaiting headers"))
```
does this error string belong to another library?
if it's internal then it's ok, but if it's not then this string can change in the next version update, and the test will break.

oshoval replied:
its internal, it a known error, that is used for long time in network libraries.

EdDev replied:
Where is this error thrown? I do not seem to find it (I grep for it and failed).

Also, the previous assertion will not give any valuable information if it fails, and it is pretty hard to follow the logic here.

- I suggest you try to find the relevant erros using the new error functionalities in Go 1.13 [1].
  (see `Examining errors with Is and As` section)
- I suggest with the previous point to make it read something like: If this is not the expected error, break from the loop. If when finishing the loop, assert against the error such that we will get the reasoning of the failure.
(you may need a helper function to return the error is the loop/retry fails eventually)

1. https://blog.golang.org/go1.13-errors

oshoval replied:
> Where is this error thrown? I do not seem to find it (I grep for it and failed).
>
Print the error and see it returns a (*url.Error)
```
client: request: &{GET https://10.244.140.74:8443/metrics HTTP/1.1 1 1 map[] <nil> <nil> 0 [] false 10.244.140.74:8443 map[] map[] <nil> map[]   <nil> <nil> <nil> 0xc0000de010} #0: Get https://10.244.140.74:8443/metrics: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```
> Also, the previous assertion will not give any valuable information if it fails, and it is pretty hard to follow the logic here.
> 
> * I suggest you try to find the relevant erros using the new error functionalities in Go 1.13 [1].
>   (see `Examining errors with Is and As` section)
> * I suggest with the previous point to make it read something like: If this is not the expected error, break from the loop. If when finishing the loop, assert against the error such that we will get the reasoning of the failure.
>   (you may need a helper function to return the error is the loop/retry fails eventually)

We already getting the reasoning on the error upon the failed Expect, i find the logic above not suitable and much harder to follow.
I tried to make it simpler, and this is the simplest and trustful we found (see discussions on 4140 if you wish also please, since this PR is part of that one).
Also please search `IsProbableEOF` vendor/k8s.io/apimachinery/pkg/util/net/http.go
The same method is used there.


> 
> 1. https://blog.golang.org/go1.13-errors

EdDev replied:
> > Where is this error thrown? I do not seem to find it (I grep for it and failed).
> 
> Print the error and see it returns a (*url.Error), and this is the string of the error which is ok, its the ip stack of some compiled lib that returns it i believe.

This contradicts the answer you gave @RamLavi when you said "its internal".
If it is not internal his comment stands and indeed it is problematic to depend on the string.

> We already getting the reasoning on the error upon the failed Expect, i find the logic above not suitable and much harder to follow.

If the error is not of the right type, no info will be provided.

I find the existing code exposing too much details which I need to decode in order to understand what is done here.
I want to read the logic without being ditracted by details, this way I will understand faster and it will help future maintenance of this code.

> 
> Also the only expected error is url.Error, so we can use Expect and fail the test instead capturing, breaking, rechecking. This is a reason for example why not to use the `As` with the logic of breaking and so on, but to keep it simple.

If it is the wrong error, you are expected to let it show at the test failure and not to silently swallow it.
`As` performs the type check for you, without the need to "manually" do it through "casting"/type assertion.

> I tried to make it simpler, and this is the simplest and trustful we found (see discussions on 4140 if you wish also please, since this PR is part of that one).
> Also please see `IsProbableEOF` vendor/k8s.io/apimachinery/pkg/util/net/http.go
> The same method is used there.

You have been suggested a different way, something more "modern".
BTW, that ref is a function that does on thing (returning if this is a specific error or not), unlike the embedded version in this PR.

oshoval replied:
```
    Expected
        <string>: dial tcp [fd10:244::8c48]:8443: connect: network is unreachable
    to contain substring
        <string>: Client.Timeout exceeded while awaiting headers
```
This what i meant that in case the string is something else we will see it, for example since there is a bug with ipv6 we see the first error happens instead the expect which is the 2nd line.

There are too much details in this comment, i honestly find it over-engineering,
i will try to first start with a function pretty similar to `IsProbableEOF` but with `As` and lets iterate from there please.

EdDev replied:
The lack of information is at this assertion: `Expect(ok).To(BeTrue())`.

Asserting on a specific string from an external library is incorrect as commented on by the first comment.
I do not know what is over-engineered in the comments, they point on specific problem of maintenance to this piece of code.

oshoval replied:
tried to take care about it, PTAL

RamLavi commented on tests/infra_test.go:
```diff
@@ -696,6 +697,9 @@ var _ = Describe("[Serial]Infrastructure", func() {
 				err := <-errors
 				if err != nil {
+					urlErr, ok := err.(*netURL.Error)
+					Expect(ok).To(BeTrue())
```
please add Expect description

oshoval replied:
in the next change, this line wont be there, so no need

RamLavi commented on tests/infra_test.go:
```diff
@@ -696,6 +697,9 @@ var _ = Describe("[Serial]Infrastructure", func() {
 				if err != nil {
+					urlErr, ok := err.(*netURL.Error)
+					Expect(ok).To(BeTrue())
+					Expect(urlErr.Err.Error()).To(ContainSubstring("Client.Timeout exceeded while awaiting headers"))
```
ditto

oshoval replied:
in case of error the string would be redundant because the error will say it already

RamLavi replied:
fair enough

RamLavi commented on tests/infra_test.go:
```diff
@@ -725,42 +729,21 @@ var _ = Describe("[Serial]Infrastructure", func() {
 				value := metrics[key]
-				Expect(value).To(BeNumerically(">=", float64(0.0)))
+				fmt.Fprintf(GinkgoWriter, "metric value was %f\n", value)
+				Expect(value).To(BeNumerically(operator, float64(0.0)))
```
ditto

oshoval replied:
didnt touch this line, just indented,
also in case of error the string would be redundant because the error will say it already

RamLavi replied:
the failed test would say "expected 2.0 to be numerically 0.0", I want it to say something more meaningful, if possible.

oshoval replied:
its not part of this PR scope, we can do it if needed on other PR please.
also since the test itself say what is the metric that is checked, this will be redundant, all the info is there, just few lines above in the log block.

RamLavi replied:
fair enough

RamLavi reviewed (changes requested):
please add Expect descriptions, as it'll make us see better where tests fail in the future

oshoval commented:
Added one more test to the DescribeTable (2nd commit).
Added the 3rd commit.

Didnt answer yet comments.

oshoval commented on tests/infra_test.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
+	netURL "net/url"
```
done

oshoval commented:
Created https://github.com/kubevirt/kubevirt/pull/4307
which is based on this PR, and adapting tests to dual stack, (see please only last commit there, rest are cherry-picks).

EdDev commented on tests/infra_test.go:
```diff
@@ -35,9 +36,11 @@ import (
 	"github.com/onsi/ginkgo/extensions/table"
 	. "github.com/onsi/gomega"
 	gomegatypes "github.com/onsi/gomega/types"
+	errs "github.com/pkg/errors"
```
Why is this needed?

oshoval replied:
`if errs.As(err, &e) {`

EdDev replied:
You do not need it, it is in the stdlib.

EdDev commented on tests/infra_test.go:
```diff
@@ -693,12 +696,21 @@ var _ = Describe("[Serial]Infrastructure", func() {
-					errorCount += 1
+					var e *neturl.Error
+					if errs.As(err, &e) {
+						Expect(e.Err.Error()).To(ContainSubstring("Client.Timeout exceeded while awaiting headers"))
```
You have not answered or resolved the issue with the text that was pointed out by @RamLavi in previous comments.
You should not assert against text from libs which you do not control. Especially when you have an alternative.

oshoval replied:
As i wrote, see please `IsProbableEOF` vendor/k8s.io/apimachinery/pkg/util/net/http.go
The same method is used there.
Also what is the alternative you suggest ? 
The only thing that is return beside this string is timeout = true in that struct, and it might be too generic to use it.

func IsProbableEOF(err error) bool {
	if err == nil {
		return false
	}
	if uerr, ok := err.(*url.Error); ok {
		err = uerr.Err
	}
	msg := err.Error()
	switch {
	case err == io.EOF:
		return true
	case msg == "http: can't write HTTP request on broken connection":
		return true
	case strings.Contains(msg, "http2: server sent GOAWAY and closed the connection"):
		return true
	case strings.Contains(msg, "connection reset by peer"):
		return true
	case strings.Contains(strings.ToLower(msg), "use of closed network connection"):
		return true
	}
	return false
}

oshoval replied:
btw the error comes from `/usr/local/go/src/net/http/client.go`
```
				err = &httpError{
					// TODO: early in cycle: s/Client.Timeout exceeded/timeout or context cancelation/
					err:     err.Error() + " (Client.Timeout exceeded while awaiting headers)",
					timeout: true,
				}
```
I believe, if it helps

EdDev replied:
I think the timeout check will suffice.
Referencing me to something that is not a timeout is not relevant and that code is unsafe due to the text check. It may worth the risk there, but here I will take the risk of the timeout check.

oshoval replied:
will make sure it catch the ipv6 error and use it if so

EdDev commented on tests/infra_test.go:
```diff
@@ -693,12 +696,21 @@ var _ = Describe("[Serial]Infrastructure", func() {
+						Expect(e.Err.Error()).To(ContainSubstring("Client.Timeout exceeded while awaiting headers"))
+						errorCount += 1
+					} else {
+						unknownErrorTypeFound = true
```
Please try to avoid this flag, it is not serving us well. We loose the error info and it does not read nicely.
I suggest you extract the whole loop to a function and return the error, then assert it.

EdDev commented on tests/infra_test.go:
```diff
@@ -783,28 +784,39 @@ var _ = Describe("[Serial]Infrastructure", func() {
-			// this tests requires at least two running VMis. To ensure this condition,
-			// the simplest way is just always run an additional VMI.
-			By("Creating another VirtualMachineInstance")
+		Context("Creating another VirtualMachineInstance", func() {
```
I do not understand the context description. it is not suppose to tell me how it implements it, but what it represents.

Context text cannot replace By text.

EdDev commented on tests/infra_test.go:
```diff
@@ -783,28 +784,39 @@ var _ = Describe("[Serial]Infrastructure", func() {
-			nodeName := startVMI(vmi)
-			Expect(nodeName).To(Equal(preferredNodeName), "Should run VMIs on the same node")
+			AfterEach(func() {
+				virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})
```
You are not checking if this fails.

oshoval replied:
it was because i didnt check its nil , so it can fail, will add both but then it wont be unconditional

EdDev replied:
Maybe I miss something, but how is unconditional relevant here? You do not have multiple resources to clean.

oshoval replied:
i dont check if vmi != nil
so it can fail , and even panic if i deref a nil ptr

EdDev replied:
So if `vmi` is `nil` and you do do `vmi.Name` you expect this do just panic? I think you interpret unconditional wrongly.

oshoval replied:
i saw we dont really care for secondary asserts, for example we dont nil the ptr.
anyway i will add it

EdDev commented on tests/infra_test.go:
```diff
@@ -693,12 +696,21 @@ var _ = Describe("[Serial]Infrastructure", func() {
 				if err != nil {
-					errorCount += 1
+					var e *neturl.Error
+					if errs.As(err, &e) {
```
`e` should be populated with the actual error, so why are we checking what it contains? `e.Err`
If you know the embedded error, you can check for it and not the wrapper.

oshoval replied:
this is what other places + the link you sent does, something like a safe cast, can you explain what you suggest?
i will try to play with it, but best if you want to explain

EdDev replied:
I think I was specific here. You look for `neturl.Error` but in practice you check the error it wraps. So I suggest you look for the wrapped error if possible instead of the wrapper.

oshoval commented:
Answered comments

EdDev commented on tests/infra_test.go:
```diff
@@ -688,14 +690,15 @@ var _ = Describe("[Serial]Infrastructure", func() {
-				err := <-errors
+				err := <-errorsChan
 				if err != nil {
+					Expect(checkError(err)).To(Succeed())
```
This reads as: I expect that the error check succeeds.
What is checked is not clear so this is not a fit.

IMO you should wrap the whole for-loop, return an error (not assert) and at line 707 assert on it (Succeed()).
e.g. `Expect(waitHttpResponse(errorsChan).To(Succeed())`

If the number of error count is important, move it into the helper function (`waitHttpResponse`) and print it there before returning.

Please see also the problem with the goroutine leak here (prev comment).

oshoval replied:
so i will just change it to BeNil , i prefer it as this if possible
i perfer the logic that iterate on the loop in the test itself, no need to send all the parameters to another level

EdDev replied:
The logic cannot be read, which makes it improper from my perspective.
I also do not understand why you think that exposing all this details is correct or better. When you cannot read it without getting into the details of the inner function, then it is not abstracting it correctly.

oshoval replied:
deleted before saw you answered, because i am changing it, even that i dont agree its not readable.

EdDev commented on tests/infra_test.go:
```diff
@@ -678,7 +680,7 @@ var _ = Describe("[Serial]Infrastructure", func() {
-			errors := make(chan error)
+			errorsChan := make(chan error)
 			for ix := 0; ix < concurrency; ix++ {
 				go func(ix int) {
```
This goroutine is unsafe as it may leak. The suggested change now will surely cause it to leak in case one of the errors are not of a timeout.

RamLavi replied:
can you elaborate on that? what leak are we talking about?

oshoval replied:
i didnt touch it, how my changes / this PR scope related to it?

the for loop which i did change, runs  `concurrency` times, each one will be either an nil err , expected timeout or unexpected timeout.

please explain better what you mean and if its not belong to the scope of this PR, lets take it to other PR please.

EdDev replied:
The changes here caused the receiver side of this groutine to change, so it is in the scope even if you did not touch these specific lines.

The channel is a blocking one (it has no buffers), therefore when `errorsChan <- err` is executed it will block until the receiving side will accept it.
But now, there receiver side will assert and may exit the loop **before** all responses have been processed. This will cause the goroutine to block, forever (i.e. leaking).

goroutines are dangerous, so if you touch it (including channels) make sure it is safe.

oshoval replied:
there is a garbage collector that will close the channel when the context is done, and then the go routine will die sending on closed channel, but i will take care of it.

EdDev replied:
> there is a garbage collector that will close the channel when the context is done, and then the go routine will die sending on closed channel, but i will take care of it.

If this is true why would you take care of it?
I am pretty sure the gc will not clean this. The gc does not close anything, it frees memory objects but does not call destruction. But if you find it is true, please share the ref.

oshoval replied:
again deleted it before you answered because gc might not delete it since the channel is still used in the go routine,
btw we can buffer the channel as one of the solution or consume them all, implemented consuming atm.

RamLavi commented on tests/infra_test.go:
```diff
@@ -909,6 +912,15 @@ var _ = Describe("[Serial]Infrastructure", func() {
 
+func checkError(err error) error {
+	var e *neturl.Error
+	if errors.As(err, &e) && e.Timeout() {
```
don't you mean to check that `&& err.Timeout()`? how can a an error you just created `e` have timeout?

EdDev replied:
e is populated by `As`.

oshoval replied:
i didnt created it, i just casted it, and it does have a timeout field, if timeout == true its the expected good error
else its a "bad error" such as `connect: network is unreachable` which we do have on ipv6

oshoval commented:
answered comments, checking it also on 3607 with injecting ipv6 to see it catches it as expected

oshoval commented on tests/infra_test.go:
```diff
@@ -975,3 +958,28 @@ func getKeysFromMetrics(metrics map[string]float64) []string {
+	errorCount := 0
+	var sampledErr error
+	for ix := 0; ix < concurrency; ix++ {
+		err := <-errorsChan
```
i might add a timer for this as well, in case the workers have panic, or we can trust them ?

oshoval commented:
Updated logic of both consumer producer, so each one would be robust to panics in the other.

RamLavi commented on tests/infra_test.go:
```diff
@@ -678,8 +680,9 @@ var _ = Describe("[Serial]Infrastructure", func() {
-			errors := make(chan error)
+			errorsChan := make(chan error, concurrency)
 			for ix := 0; ix < concurrency; ix++ {
+				defer GinkgoRecover()
```
shouldn't the defer exist inside the goroutine? see [example](https://onsi.github.io/ginkgo/#marking-specs-as-failed)

oshoval replied:
totally

RamLavi commented on tests/infra_test.go:
```diff
@@ -688,19 +691,12 @@ var _ = Describe("[Serial]Infrastructure", func() {
-
-			fmt.Fprintf(GinkgoWriter, "client: total errors #%d\n", errorCount) // troubleshooting helper
+			errorCount, err := waitHttpResponse(errorsChan, concurrency)
+			Expect(err).To(BeNil())
```
perhaps using this is more readable?
```suggestion
			Expect(err).ToNot(HaveOccurred())
```

oshoval commented:
Answered Ram's comments

EdDev commented on tests/infra_test.go:
```diff
@@ -768,28 +769,42 @@ var _ = Describe("[Serial]Infrastructure", func() {
-			// this tests requires at least two running VMis. To ensure this condition,
-			// the simplest way is just always run an additional VMI.
-			By("Creating another VirtualMachineInstance")
+		Context("when creating another VirtualMachineInstance", func() {
```
I remember that I commented about the context description here. The context name should express the what and not the how. Try to build a sentence with the context/describe that wraps and the test below it and check if it sounds informative.

I think we agreed to keep the current status as is in non-network tests. In this case, I see `tests.BeforeTestCleanup()` is used, so is there any other reason we explicitly have to add this change?

oshoval replied:
i added it because the test itself create a vm and doesn't delete it, and once we adapt it to dual stack we must delete it (see #4307) because the metrics in this specific test count vms, and if we dont delete it the count wont be right,
also since this test create a vm it should delete it, and BeforeTestCleanup wont help in this situation.

oshoval replied:
Updated the when, thanks for the tips

oshoval commented on tests/infra_test.go:
```diff
@@ -975,3 +971,34 @@ func getKeysFromMetrics(metrics map[string]float64) []string {
+	for ix := 0; ix < concurrency; ix++ {
+		select {
+		case err = <-errorsChan:
+			if err != nil {
+				err = checkError(err)
+				if err != nil {
+					break
+				}
+				expectedErrorsCount += 1
+			}
```
i count it only if its an expected error, in case its unexpected error its a fatal and we can break out
i dont need to create a list of error pools, just to return the first offender
we dont need to count timeout ones, i will also remove it because the comment of 
https://github.com/kubevirt/kubevirt/pull/4306#discussion_r503234748
even that i think its good enough imo.

oshoval replied:
updated the func as we spoke

oshoval commented on tests/infra_test.go:
```diff
@@ -678,29 +680,23 @@ var _ = Describe("[Serial]Infrastructure", func() {
-			fmt.Fprintf(GinkgoWriter, "client: total errors #%d\n", errorCount) // troubleshooting helper
+			errorCount, err := waitHttpResponse(errorsChan, concurrency)
+			Expect(err).ToNot(HaveOccurred(), "Should succeed receiving HTTP responses without unexpected errors")
 			Expect(errorCount).To(BeNumerically(">", 0))
```
if there arent errors at all, we still need to know about it, as this test check throttling.

oshoval commented on tests/infra_test.go:
```diff
@@ -678,29 +680,23 @@ var _ = Describe("[Serial]Infrastructure", func() {
+			errorsChan := make(chan error, concurrency)
 			for ix := 0; ix < concurrency; ix++ {
 				go func(ix int) {
+					defer GinkgoRecover()
```
as i wrote in other thread here, indeed there arent fail here, but it might be a good practice to have it on go routines?

oshoval commented on tests/infra_test.go:
```diff
@@ -975,3 +971,34 @@ func getKeysFromMetrics(metrics map[string]float64) []string {
+				}
+				expectedErrorsCount += 1
+			}
+		case <-time.After(5 * time.Second):
```
the serial time will happen at most one time in case of a failure, the select will catch the first one of those events, so it doesnt had time, unless a fatal which shouldnt occur happen and then the test fails (for example hanged producer).
anyway i will remove it, even that i dont agree that its not enough.
it makes the consumer  / producer each one protected on its own without dependency.

oshoval replied:
Removed the timeout, as its also break the draining (one would not been consumed) so if we remove the channel buffer we must remove the timeout as well.

oshoval commented:
/retest

oshoval commented:
Addressed comments, thanks
testing it locally meanwhile

EdDev commented on tests/infra_test.go:
```diff
@@ -975,3 +969,32 @@ func getKeysFromMetrics(metrics map[string]float64) []string {
+	}
+}
+
+func validatedHTTPResponses(errorsChan chan error, concurrency int) error {
```
Please add a doctext here and explain what it validates, it took me some time to understand this after understanding a day before.

EdDev commented on tests/infra_test.go:
```diff
@@ -975,3 +969,32 @@ func getKeysFromMetrics(metrics map[string]float64) []string {
+	}
+
+	if sampledError == nil && expectedErrorsCount == 0 {
+		sampledError = fmt.Errorf("timeout errors were expected due to throttling")
```
Either return here directly or change the variable name to be generic... it is not a sampleError anymore after this assignment and it is confusing.

EdDev commented on tests/infra_test.go:
```diff
@@ -975,3 +969,32 @@ func getKeysFromMetrics(metrics map[string]float64) []string {
 	return keys
 }
+
+func checkError(err error) error {
+	var e *neturl.Error
+	if errors.As(err, &e) && e.Timeout() {
+		return nil
+	} else {
+		return err
+	}
+}
```
I do not think you need this whole function, you just need the condition from it...
`errors.As(err, &e) && e.Timeout()`

So you can change it to `isTimeoutError`and simplify it or embed it in the caller one.

oshoval commented:
Updated according comments

EdDev reviewed (approved):
Thank you!

oshoval commented:
As the last commit here use AfterEach cleanup, in case it would fail, it wont collect info, 

possible approaches:
1. Unadapt this specific test 4146 to a table for dual stack, use a for loop instead (on 4307).
2. Refactor 4146 to be in its own context and use CleanAll function (pretty big change since lots of functions are needed, already started, but if we have a faster decision it would be great).
3. Add a variable to test 4146 that will count how many new machines this test created and then we create dependency in this table only, Edy less like it, it creates a simple dependency.
4. Keep the PR as its now, just in case of an error it wont collect the machines, but they are actually not important there because the test just count the vms in the metrics.
5. We have enough metrics checked, we can even dont adapt this one to ipv6, its not the main issue of this file, and we already did on most. 

Please suggest what you prefer

Thanks

AlonaKaplan commented on tests/infra_test.go:
```diff
@@ -688,20 +690,12 @@ var _ = Describe("[Serial]Infrastructure", func() {
 					}
-					errors <- err
+					errorsChan <- err
 				}(ix)
```
I would move the error treatment (`validatedHTTPResponses`) to here. So if we get a timeout error we can break the loop.

AlonaKaplan commented on tests/infra_test.go:
```diff
@@ -975,3 +969,30 @@ func getKeysFromMetrics(metrics map[string]float64) []string {
+// there were no errors at all.
+func validatedHTTPResponses(errorsChan chan error, concurrency int) error {
+	var expectedErrorsCount int = 0
+	var sampledError error
```
Maybe s/sampledError/unexpectedError?

oshoval replied:
Done

oshoval commented:
Addressed comments.
Implemented option 6 of above, reverted the latest commit so it will be still shown for the discussion, at the end i will squash it.
Locally also implemented option 2 which is overkill, didn't include it, because option 6 imho is the best.

AlonaKaplan commented on tests/infra_test.go:
```diff
@@ -767,25 +767,13 @@ var _ = Describe("[Serial]Infrastructure", func() {
 		})
 
 		It("[test_id:4146]should include VMI phase metrics for few running VMs", func() {
-			// this tests requires at least two running VMis. To ensure this condition,
```
Who creates the second vmi?

oshoval replied:
its already there always in the BeforeAll

https://github.com/kubevirt/kubevirt/pull/4306/files#diff-c95f11746affb831bc43f837f8be65ad753336c090e3b181531ec00bf7636d7dL529-L531
```
			nodeName := prepareVMIForTests("")				nodeName := prepareVMIForTests("")
			// any node is fine, we don't really care, as long as we run all VMIs on it.				// any node is fine, we don't really care, as long as we run all VMIs on it.
			prepareVMIForTests(nodeName)
```

AlonaKaplan replied:
thanks

AlonaKaplan commented:
The PR is fine. Please remove commits 3 and 4 and I will approve it.

oshoval commented:
> The PR is fine. Please remove commits 3 and 4 and I will approve it.

Done thanks

AlonaKaplan commented:
/approve

oshoval commented:
/retest

RamLavi commented:
/lgtm
